File: packages/skill-eval/src/report/json-reporter.ts
import fs from "node:fs";
import type { ResolvedConfig } from "../config/schema.js";
import type { OrchestratorResult } from "../runner/orchestrator.js";
import type { BenchmarkFile } from "../schemas/benchmark.js";

export function writeBenchmarkJson(
	result: OrchestratorResult,
	config: ResolvedConfig,
	meta: {
		skillName: string;
		skillPath: string;
		evalsRun: number[];
	},
	outputPath: string,
) {
	const benchmark: BenchmarkFile = {
		metadata: {
			skill_name: meta.skillName,
			skill_path: meta.skillPath,
			timestamp: new Date().toISOString(),
			evals_run: meta.evalsRun,
			runs_per_provider: config.settings.runsPerProvider,
			providers: config.providers.map((p) => ({ name: p.name, model: p.model })),
			grader: { name: config.grader.provider, model: config.grader.model },
		},
		runs: result.runs,
		provider_summary: result.providerSummary,
		notes: [],
	};

	fs.writeFileSync(outputPath, `${JSON.stringify(benchmark, null, 2)}\n`);
	return outputPath;
}

File: packages/skillet-eval-python/src/skill_eval/report/json_reporter.py
from __future__ import annotations

from datetime import datetime
from pathlib import Path

from skill_eval.config.schema import ResolvedConfig
from skill_eval.schemas.benchmark import (
    BenchmarkFile,
    BenchmarkMetadata,
    ProviderInfo,
    ProviderSummary,
)


def write_benchmark_json(
    result: dict,
    config: ResolvedConfig,
    meta: dict,
    output_path: str,
) -> str:
    provider_summary_raw = result.get("provider_summary") or result.get("providerSummary") or {}
    provider_summary = {
        k: ProviderSummary.model_validate(v) for k, v in provider_summary_raw.items()
    }

    benchmark = BenchmarkFile(
        metadata=BenchmarkMetadata(
            skill_name=meta["skill_name"],
            skill_path=meta["skill_path"],
            timestamp=datetime.now().isoformat(),
            evals_run=meta["evals_run"],
            runs_per_provider=config.settings.runs_per_provider,
            providers=[
                ProviderInfo(name=p.name, model=p.model) for p in config.providers
            ],
            grader=ProviderInfo(
                name=config.grader.provider,
                model=config.grader.model,
            ),
        ),
        runs=result["runs"],
        provider_summary=provider_summary,
        notes=[],
    )

    Path(output_path).write_text(
        benchmark.model_dump_json(indent=2) + "\n",
        encoding="utf-8",
    )
    return output_path
